service/src/test/java/bio/terra/workspace/service/resource/controlled/cloud/aws/s3StorageFolder/AwsS3StorageFolderStepTest.java

```diff
@@ -0,0 +1,353 @@
+
+import bio.terra.common.exception.ApiException;
+import bio.terra.common.exception.ConflictException;
+import bio.terra.common.exception.UnauthorizedException;
+import bio.terra.stairway.FlightContext;
+import bio.terra.stairway.StepResult;
+import bio.terra.stairway.StepStatus;
+import bio.terra.workspace.common.BaseAwsUnitTest;
+import bio.terra.workspace.common.exception.InternalLogicException;
+import bio.terra.workspace.common.fixtures.ControlledAwsResourceFixtures;
+import bio.terra.workspace.common.fixtures.ControlledResourceFixtures;
+import bio.terra.workspace.common.fixtures.WorkspaceFixtures;
+import bio.terra.workspace.common.utils.AwsUtils;
+import bio.terra.workspace.common.utils.MockMvcUtils;
+import bio.terra.workspace.service.iam.AuthenticatedUserRequest;
+import bio.terra.workspace.service.iam.SamService;
+import bio.terra.workspace.service.workspace.AwsCloudContextService;
+import java.util.Optional;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.TestInstance;
+import org.junit.jupiter.api.TestInstance.Lifecycle;
+import org.mockito.Mock;
+import org.mockito.MockedStatic;
+import org.mockito.Mockito;
+import software.amazon.awssdk.awscore.exception.AwsServiceException;
+import software.amazon.awssdk.core.sync.RequestBody;
+import software.amazon.awssdk.services.s3.S3Client;
+import software.amazon.awssdk.services.s3.model.DeleteObjectsRequest;
+import software.amazon.awssdk.services.s3.model.DeleteObjectsResponse;
+import software.amazon.awssdk.services.s3.model.ListObjectsV2Request;
+import software.amazon.awssdk.services.s3.model.ListObjectsV2Response;
+import software.amazon.awssdk.services.s3.model.PutObjectRequest;
+import software.amazon.awssdk.services.s3.model.PutObjectResponse;
+import software.amazon.awssdk.services.s3.model.S3Error;
+import software.amazon.awssdk.services.s3.model.S3Exception;
+import software.amazon.awssdk.services.s3.model.S3Object;
+
+@TestInstance(Lifecycle.PER_CLASS)
+public class AwsS3StorageFolderStepTest extends BaseAwsUnitTest {
+
+  @Mock private FlightContext mockFlightContext;
+  @Mock private AwsCloudContextService mockAwsCloudContextService;
+  @Mock private SamService mockSamService;
+  @Mock private S3Client mockS3Client;
+
+  private MockedStatic<AwsUtils> mockAwsUtils;
+  private ControlledAwsS3StorageFolderResource s3FolderResource;
+  private static final AwsServiceException s3Exception1 =
+      S3Exception.builder().message("not authorized to perform").build();
+
+  @BeforeAll
+  public void init() {
+    mockAwsUtils = Mockito.mockStatic(AwsUtils.class);
+    s3FolderResource =
+        ControlledAwsResourceFixtures.makeDefaultAwsS3StorageFolderResource(
+            ControlledResourceFixtures.WORKSPACE_ID);
+  }
+
+  @BeforeEach
+  public void setup() {
+    when(mockFlightContext.getResult())
+        .thenReturn(new StepResult(StepStatus.STEP_RESULT_FAILURE_FATAL));
```

Comment:
Out of curiosity, why is this always mocked to return a failure?

Reply:
- A flight in `do direction` is still running and doesn't have a result, hence the getResult() will not have value in do().
- undo() is invoked when a flight fails and getResult() should reflect that a flight failed, hence always returning error